Tutorial/brusselator/analysis.cpp

```diff
@@ -0,0 +1,190 @@
+ * Writes the variables and their computed norm to an ADIOS2 file.
+ *
+ * Kshitij Mehta
+ *
+ * @TODO:
+ *      - Error checks. What is vector resizing returns an out-of-memory error? Must handle it
+ *      - Turn ADIOS2 Debug to ON
+ */
+#include <iostream>
+#include <stdexcept>
+#include <cstdint>
+#include <cmath>
+#include "adios2.h"
+
+/*
+ * Function to compute the norm of an array
+ */
+std::vector<double> compute_norm (std::vector<double> real_part, std::vector<double> imag_part, int dims) {
+    int i;
+    std::vector<double> norm;
+    int comm_size;
+    MPI_Comm_size(MPI_COMM_WORLD, &comm_size);
+
+    norm.resize(dims);
+    for (i=0; i<dims; i++)
+        norm[i] = sqrt( real_part[i]*real_part[i] + imag_part[i]*imag_part[i] );
+
+    return norm;
+}
+
+/*
+ * MAIN
+ */
+int main(int argc, char *argv[])
+{
+    MPI_Init(&argc, &argv);
+    int rank, comm_size;
+    
+    int u_global_size, v_global_size;
+    int u_local_size, v_local_size;
+    
+    bool firstStep = true;
+
+    MPI_Comm_rank (MPI_COMM_WORLD, &rank);
+    MPI_Comm_size (MPI_COMM_WORLD, &comm_size);
+
+    std::vector<long unsigned int> shape_u_real;
+    std::vector<long unsigned int> shape_u_imag;
+    std::vector<long unsigned int> shape_v_real;
+    std::vector<long unsigned int> shape_v_imag;
+    
+    std::vector<double> u_real_data;
+    std::vector<double> u_imag_data;
+    std::vector<double> v_real_data;
+    std::vector<double> v_imag_data;
+
+    std::vector<double> norm_u;
+    std::vector<double> norm_v;
+    
+    // adios2 variable declarations
+    adios2::Variable<double> var_u_real_in, var_u_imag_in, var_v_real_in, var_v_imag_in;
+    adios2::Variable<double> var_u_norm, var_v_norm;
+    adios2::Variable<double> var_u_real_out, var_u_imag_out, var_v_real_out, var_v_imag_out;
+
+    // adios2 io object and engine init
+    adios2::ADIOS ad ("analysis_adios2.xml", MPI_COMM_WORLD, adios2::DebugOFF);
+
+    // IO object and engine for reading
+    adios2::IO reader_io = ad.DeclareIO("analysis_reader");
+    adios2::Engine reader_engine = reader_io.Open("./data/brusselator.bp", adios2::Mode::Read, MPI_COMM_WORLD);
+
+    // IO object and engine for writing
+    adios2::IO writer_io = ad.DeclareIO("analysis_writer");
+    adios2::Engine writer_engine = writer_io.Open("./data/analysis.bp", adios2::Mode::Write, MPI_COMM_WORLD);
+
+    // read data per timestep
+    while(true) {
+
+        // Begin step
+        adios2::StepStatus read_status  = reader_engine.BeginStep (adios2::StepMode::NextAvailable, 0.0f);
+        if (read_status != adios2::StepStatus::OK)
+            break;
+
+        if (firstStep) {
+            // Inquire variable and set the selection at the first step only
+            // This assumes that the variable dimensions do not change across timesteps
+
+            // Inquire variable
+            var_u_real_in = reader_io.InquireVariable<double>("u_real");
```

Comment:
var_u_real_in will disappear at the next BeginStep(). So does the selection. Does the read actually work in the second and further steps?

Reply:
var_u_real_in was declared outside the scope, so it won't lose scope after the first step. I am calling InquireVariable, SetSelection, and DeclareVariable only for the first step, as the shape of the variables stays constant across all timesteps.